include/seastar/core/semaphore.hh

```diff
@@ -165,37 +166,94 @@ private:
         state(basic_semaphore& sem_, ssize_t count_) noexcept : sem(&sem_), count(count_) {}
     };
     lw_shared_ptr<state> _state;
-    struct entry {
-        promise<> pr;
+
+    class basic_waiter : public boost::intrusive::list_base_hook<boost::intrusive::link_mode<boost::intrusive::auto_unlink>> {
+    public:
+        basic_semaphore* sem;
         size_t nr;
-        std::optional<abort_on_expiry<clock>> timer;
-        entry(promise<>&& pr_, size_t nr_) noexcept : pr(std::move(pr_)), nr(nr_) {}
-    };
-    struct expiry_handler {
-        basic_semaphore& sem;
-        void operator()(entry& e) noexcept {
-            if (e.timer) {
+
+    private:
+        std::variant<optimized_optional<abort_source::subscription>, timer> _aborter;
+
+    public:
+        basic_waiter() noexcept : sem(nullptr), nr(0) {}
+        basic_waiter(basic_semaphore& sem_, size_t nr_) noexcept
+            : sem(&sem_), nr(nr_)
+        {
+            sem->add_waiter(*this);
+        }
+        basic_waiter(basic_semaphore& sem_, size_t nr_, time_point timeout) noexcept
+            : sem(&sem_), nr(nr_)
+        {
+            set_timeout(timeout);
+            sem->add_waiter(*this);
+        }
+        basic_waiter(basic_semaphore& sem_, size_t nr_, abort_source& as) noexcept
+            : sem(&sem_), nr(nr_)
+        {
+            subscribe(as);
+            sem->add_waiter(*this);
+        }
+        virtual ~basic_waiter() = default;
+
+        void init(basic_semaphore& sem_, size_t nr_) noexcept {
+            sem = &sem_;
+            nr = nr_;
+            sem->add_waiter(*this);
+        }
+
+        void subscribe(abort_source& as) noexcept {
+            _aborter.template emplace<optimized_optional<abort_source::subscription>>(as.subscribe([this] (const std::optional<std::exception_ptr>& opt_ex) noexcept {
+                std::exception_ptr ex;
                 try {
-                    e.pr.set_exception(sem.timeout());
+                    if (opt_ex) {
+                        ex = *opt_ex;
+                    } else if (sem) {
+                        if (sem->_state->ex) {
+                            ex = sem->_state->ex;
+                        } else if constexpr (internal::has_aborted<ExceptionFactory>::value) {
+                            ex = std::make_exception_ptr(sem->aborted());
+                        }
+                    }
                 } catch (...) {
-                    e.pr.set_exception(semaphore_timed_out());
```

Comment:
created https://github.com/scylladb/scylladb/pull/12759 to prepare for this change.

Reply:
right.  but changing the ExceptionFactory methods to be noexcept is an api change.
I'm not sure we want to change without bumping the api version since it could break users derived implementations.

Reply:
yeah, makes sense to me. i checked redpanda and ceph/crimson's code. they are not using this feature. but yeah. that's an API change.